api/trace/b3_propagator.go

```diff
@@ -34,185 +34,295 @@ const (
 	sc := SpanFromContext(ctx).SpanContext()
-	if !sc.IsValid() {
-		return
+
+	if b3.supports(SingleHeader) {
+		header := []string{}
+		if sc.TraceID.IsValid() && sc.SpanID.IsValid() {
+			header = append(header, sc.TraceID.String(), sc.SpanID.String())
+		}
+
+		if sc.isDebug() {
+			header = append(header, "d")
+		} else if !sc.isDeferred() {
+			if sc.IsSampled() {
+				header = append(header, "1")
+			} else {
+				header = append(header, "0")
+			}
+		}
+
+		supplier.Set(B3SingleHeader, strings.Join(header, "-"))
 	}
-	if b3.SingleHeader {
-		sampled := sc.TraceFlags & FlagsSampled
-		supplier.Set(B3SingleHeader,
-			fmt.Sprintf("%s-%s-%.1d", sc.TraceID, sc.SpanID, sampled))
-	} else {
-		supplier.Set(B3TraceIDHeader, sc.TraceID.String())
-		supplier.Set(B3SpanIDHeader, sc.SpanID.String())
 
-		var sampled string
-		if sc.IsSampled() {
-			sampled = "1"
-		} else {
-			sampled = "0"
+	if b3.supports(MultipleHeader) || b3.InjectEncoding == 0 {
+		if sc.TraceID.IsValid() && sc.SpanID.IsValid() {
+			supplier.Set(B3TraceIDHeader, sc.TraceID.String())
+			supplier.Set(B3SpanIDHeader, sc.SpanID.String())
+		}
+
+		if sc.isDebug() {
+			// Since Debug implies deferred, don't also send "X-B3-Sampled".
+			supplier.Set(B3DebugFlagHeader, "1")
+		} else if !sc.isDeferred() {
+			if sc.IsSampled() {
+				supplier.Set(B3SampledHeader, "1")
+			} else {
+				supplier.Set(B3SampledHeader, "0")
+			}
 		}
-		supplier.Set(B3SampledHeader, sampled)
 	}
 }
 
-// Extract retrieves B3 Headers from the supplier
+// Extract extracts a context from the supplier if it contains B3 headers.
 func (b3 B3) Extract(ctx context.Context, supplier propagation.HTTPSupplier) context.Context {
-	var sc SpanContext
-	if b3.SingleHeader {
-		sc = b3.extractSingleHeader(supplier)
-	} else {
-		sc = b3.extract(supplier)
+	var (
+		sc  SpanContext
+		err error
+	)
+
+	// Default to Single Header if a valid value exists.
+	if h := supplier.Get(B3SingleHeader); h != "" {
+		sc, err = extractSingle(h)
+		if err == nil && sc.IsValid() {
+			return ContextWithRemoteSpanContext(ctx, sc)
+		}
+		// The Single Header value was invalid, fallback to Multiple Header.
 	}
-	if !sc.IsValid() {
+
+	var (
+		traceID      = supplier.Get(B3TraceIDHeader)
+		spanID       = supplier.Get(B3SpanIDHeader)
+		parentSpanID = supplier.Get(B3ParentSpanIDHeader)
+		sampled      = supplier.Get(B3SampledHeader)
+		debugFlag    = supplier.Get(B3DebugFlagHeader)
+	)
+	sc, err = extractMultiple(traceID, spanID, parentSpanID, sampled, debugFlag)
+	if err != nil || !sc.IsValid() {
 		return ctx
 	}
 	return ContextWithRemoteSpanContext(ctx, sc)
 }
 
-func fixB3TID(in string) string {
-	if len(in) == 16 {
-		in = b3TraceIDPadding + in
+func (b3 B3) GetAllKeys() []string {
+	header := []string{}
+	if b3.supports(SingleHeader) {
+		header = append(header, B3SingleHeader)
 	}
-	return in
+	if b3.supports(MultipleHeader) || b3.InjectEncoding == 0 {
```

Comment:
```suggestion
	if b3.supports(MultipleHeader) || b3.InjectEncoding == unspecified {
```

An `unspecified` constant here would be good to clarify intent.